apps/shared/avifpng.c

```diff
@@ -503,8 +503,40 @@ avifBool avifPNGWrite(const char * outputFilename, const avifImage * avif, uint3
     }
 
     png_set_IHDR(png, info, avif->width, avif->height, rgbDepth, colorType, PNG_INTERLACE_NONE, PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);
-    if (avif->icc.data && (avif->icc.size > 0)) {
+
+    const avifColorPrimaries primaries = avif->colorPrimaries;
+    const avifTransferCharacteristics transfer = avif->transferCharacteristics;
+    const avifBool hasIcc = avif->icc.data && (avif->icc.size > 0);
+    const avifBool hasNonSrgbPrimaries = (primaries != AVIF_COLOR_PRIMARIES_UNKNOWN &&
+                                          primaries != AVIF_COLOR_PRIMARIES_UNSPECIFIED && primaries != AVIF_COLOR_PRIMARIES_BT709);
+    const avifBool hasNonSrgbTransfer =
+        (transfer != AVIF_TRANSFER_CHARACTERISTICS_UNKNOWN && transfer != AVIF_TRANSFER_CHARACTERISTICS_UNSPECIFIED &&
+         transfer != AVIF_TRANSFER_CHARACTERISTICS_SRGB);
+
+    avifBool writeCicp = AVIF_FALSE;
+    if (hasIcc) {
         png_set_iCCP(png, info, "libavif", 0, avif->icc.data, (png_uint_32)avif->icc.size);
+    } else if (hasNonSrgbPrimaries || hasNonSrgbTransfer) {
+        float primariesCoords[8];
+        avifColorPrimariesGetValues(avif->colorPrimaries, primariesCoords);
+        png_set_cHRM(png,
+                     info,
+                     primariesCoords[6],
+                     primariesCoords[7],
+                     primariesCoords[0],
+                     primariesCoords[1],
+                     primariesCoords[2],
+                     primariesCoords[3],
+                     primariesCoords[4],
+                     primariesCoords[5]);
+        float gamma;
+        if (avifTransferCharacteristicsGamma(avif->transferCharacteristics, &gamma)) {
+            png_set_gAMA(png, info, 1.0f / gamma);
+        } else {
+            // The transfer characteristics cannot be represented with a simple gAma chunk.
+            // A cICP chunk is needed, but might not be understood by all viewers since it's new.
+            writeCicp = AVIF_TRUE;
+        }
```

Comment:
Just an idea: rather than making these exclusive, it could be ok to write everything that is available, and let the PNG readers parse the color encoding chunks in order of priority in the spec and their level of support?

Reply:
I'm adding the cICP chunk only if it's needed as a small size optimization, but I guess we don't really care, so I could simplify this. However, even when cICP is added, I *am* still adding cHR+gAMA as a fallback as you said, so it's not exclusive.

We can't easily add other chunks in case ICC is present, because we would need to parse the ICC to get the corresponding primaries/gamma. I don't think we should use avif->colorPrimaries and avif->transferCharacteristics in this case, as they are supposed to be irrelevant when an ICC is present as explained in https://github.com/AOMediaCodec/libavif/wiki/CICP

As an aside also note that the priority between CICP and ICC is different in the two formats: ICC has priority in avif, while CICP has priority in png.

Another chunk that could also be added is the sRGB chunk, but I think all viewers just assume sRGB if no color info is present? so I don't think it's really necessary but could be added if colorPrimaries is 1 and transferCharacteristics is 13.

Reply:
I'm not sure we're overly concerned about png size, so simplifying things where you can might make sense.

Reply:
Fair enough. I changed the logic to write as many chunks as we can, including the sRGB chunk.